File: dpb-mcp-go/pkg/analyzer/tracker.go
// Package analyzer - Dependency Tracker for Go implementation
// Tracks dependency changes with timestamps for reverting/replacing non-compliant dependencies

package analyzer

import (
	"crypto/sha256"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"
)

const TrackerFile = ".dpb-dependency-tracker.json"

// DependencySnapshot represents a point-in-time view of all dependencies
type DependencySnapshot struct {
	Timestamp    string                   `json:"timestamp"`
	Checksum     string                   `json:"checksum"`
	Dependencies []TrackedDependency      `json:"dependencies"`
	Metadata     SnapshotMetadata         `json:"metadata"`
}

// TrackedDependency represents a single dependency with tracking info
type TrackedDependency struct {
	Name           string `json:"name"`
	Version        string `json:"version"`
	Type           string `json:"type"` // "production" or "development"
	AddedAt        string `json:"addedAt,omitempty"`
	UpdatedAt      string `json:"updatedAt,omitempty"`
	License        string `json:"license,omitempty"`
	SecurityStatus string `json:"securityStatus,omitempty"`
}

// SnapshotMetadata contains additional info about the snapshot
type SnapshotMetadata struct {
	RepoPath       string `json:"repoPath"`
	PackageManager string `json:"packageManager"`
	TotalCount     int    `json:"totalCount"`
}

// DependencyChange represents a change between snapshots
type DependencyChange struct {
	Type       string `json:"type"` // "added", "removed", "updated"
	Name       string `json:"name"`
	OldVersion string `json:"oldVersion,omitempty"`
	NewVersion string `json:"newVersion,omitempty"`
	Timestamp  string `json:"timestamp"`
	Reason     string `json:"reason,omitempty"`
}

// ComplianceIssue represents a dependency compliance problem
type ComplianceIssue struct {
	Dependency       string `json:"dependency"`
	Version          string `json:"version"`
	Issue            string `json:"issue"` // "license", "security", "outdated", "deprecated"
	Severity         string `json:"severity"` // "critical", "high", "medium", "low"
	Description      string `json:"description"`
	Recommendation   string `json:"recommendation"`
	AutoFixAvailable bool   `json:"autoFixAvailable"`
}

// DependencyHistory contains current snapshot and categorized dependencies
type DependencyHistory struct {
	CurrentSnapshot DependencySnapshot  `json:"currentSnapshot"`
	RecentlyAdded   []TrackedDependency `json:"recentlyAdded"`
	RecentlyUpdated []TrackedDependency `json:"recentlyUpdated"`
	Stale           []TrackedDependency `json:"stale"`
}

// CreateDependencySnapshot creates a new snapshot of all dependencies
func CreateDependencySnapshot(repoPath string) (*DependencySnapshot, error) {
	deps, err := AnalyzeDependenciesRaw(repoPath)
	if err != nil {
		return nil, err
	}

	now := time.Now().UTC().Format(time.RFC3339)
	
	// Load existing tracker to preserve timestamps
	existing, _ := LoadTracker(repoPath)
	existingDeps := make(map[string]TrackedDependency)
	if existing != nil {
		for _, d := range existing.Dependencies {
			existingDeps[d.Name] = d
		}
	}

	var tracked []TrackedDependency

	// Process production dependencies
	for _, pkg := range deps.Tree {
		if pkg.Type == "production" {
			existing, found := existingDeps[pkg.Name]
			td := TrackedDependency{
				Name:           pkg.Name,
				Version:        pkg.Version,
				Type:           "production",
				License:        pkg.License,
				SecurityStatus: "unknown",
			}
			if found {
				td.AddedAt = existing.AddedAt
				if existing.Version != pkg.Version {
					td.UpdatedAt = now
				} else {
					td.UpdatedAt = existing.UpdatedAt
				}
			} else {
				td.AddedAt = now
				td.UpdatedAt = now
			}
			tracked = append(tracked, td)
		}
	}

	// Process dev dependencies
	for _, pkg := range deps.Tree {
		if pkg.Type == "development" {
			existing, found := existingDeps[pkg.Name]
			td := TrackedDependency{
				Name:           pkg.Name,
				Version:        pkg.Version,
				Type:           "development",
				License:        pkg.License,
				SecurityStatus: "unknown",
			}
			if found {
				td.AddedAt = existing.AddedAt
				if existing.Version != pkg.Version {
					td.UpdatedAt = now
				} else {
					td.UpdatedAt = existing.UpdatedAt
				}
			} else {
				td.AddedAt = now
				td.UpdatedAt = now
			}
			tracked = append(tracked, td)
		}
	}

	// Calculate checksum
	var names []string
	for _, d := range tracked {
		names = append(names, fmt.Sprintf("%s@%s", d.Name, d.Version))
	}
	sort.Strings(names)
	hash := sha256.Sum256([]byte(fmt.Sprintf("%v", names)))
	checksum := fmt.Sprintf("%x", hash)[:16]

	snapshot := &DependencySnapshot{
		Timestamp:    now,
		Checksum:     checksum,
		Dependencies: tracked,
		Metadata: SnapshotMetadata{
			RepoPath:       repoPath,
			PackageManager: "composer",
			TotalCount:     len(tracked),
		},
	}

	return snapshot, nil
}

// LoadTracker loads existing tracker data from file
func LoadTracker(repoPath string) (*DependencySnapshot, error) {
	trackerPath := filepath.Join(repoPath, TrackerFile)
	data, err := os.ReadFile(trackerPath)
	if err != nil {
		return nil, err
	}

	var snapshot DependencySnapshot
	if err := json.Unmarshal(data, &snapshot); err != nil {
		return nil, err
	}

	return &snapshot, nil
}

// SaveSnapshot saves a snapshot to the tracker file
func SaveSnapshot(repoPath string, snapshot *DependencySnapshot) error {
	trackerPath := filepath.Join(repoPath, TrackerFile)
	data, err := json.MarshalIndent(snapshot, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(trackerPath, data, 0644)
}

// CompareSnapshots returns the differences between two snapshots
func CompareSnapshots(oldSnapshot, newSnapshot *DependencySnapshot) []DependencyChange {
	var changes []DependencyChange

	oldDeps := make(map[string]TrackedDependency)
	for _, d := range oldSnapshot.Dependencies {
		oldDeps[d.Name] = d
	}

	newDeps := make(map[string]TrackedDependency)
	for _, d := range newSnapshot.Dependencies {
		newDeps[d.Name] = d
	}

	// Find added and updated
	for name, newDep := range newDeps {
		if oldDep, found := oldDeps[name]; !found {
			changes = append(changes, DependencyChange{
				Type:       "added",
				Name:       name,
				NewVersion: newDep.Version,
				Timestamp:  newSnapshot.Timestamp,
			})
		} else if oldDep.Version != newDep.Version {
			changes = append(changes, DependencyChange{
				Type:       "updated",
				Name:       name,
				OldVersion: oldDep.Version,
				NewVersion: newDep.Version,
				Timestamp:  newSnapshot.Timestamp,
			})
		}
	}

	// Find removed
	for name, oldDep := range oldDeps {
		if _, found := newDeps[name]; !found {
			changes = append(changes, DependencyChange{
				Type:       "removed",
				Name:       name,
				OldVersion: oldDep.Version,
				Timestamp:  newSnapshot.Timestamp,
			})
		}
	}

	return changes
}

// GetDependencyHistory returns categorized dependency information
func GetDependencyHistory(repoPath string) (*DependencyHistory, error) {
	snapshot, err := CreateDependencySnapshot(repoPath)
	if err != nil {
		return nil, err
	}

	now := time.Now()
	thirtyDaysAgo := now.AddDate(0, 0, -30)
	oneYearAgo := now.AddDate(-1, 0, 0)

	var recentlyAdded, recentlyUpdated, stale []TrackedDependency

	for _, dep := range snapshot.Dependencies {
		if dep.AddedAt != "" {
			addedTime, _ := time.Parse(time.RFC3339, dep.AddedAt)
			if addedTime.After(thirtyDaysAgo) {
				recentlyAdded = append(recentlyAdded, dep)
			}
		}

		if dep.UpdatedAt != "" && dep.UpdatedAt != dep.AddedAt {
			updatedTime, _ := time.Parse(time.RFC3339, dep.UpdatedAt)
			if updatedTime.After(thirtyDaysAgo) {
				recentlyUpdated = append(recentlyUpdated, dep)
			}
			if updatedTime.Before(oneYearAgo) {
				stale = append(stale, dep)
			}
		}
	}

	return &DependencyHistory{
		CurrentSnapshot: *snapshot,
		RecentlyAdded:   recentlyAdded,
		RecentlyUpdated: recentlyUpdated,
		Stale:           stale,
	}, nil
}

// CheckCompliance checks dependencies for compliance issues
func CheckCompliance(repoPath string) ([]ComplianceIssue, error) {
	snapshot, err := CreateDependencySnapshot(repoPath)
	if err != nil {
		return nil, err
	}

	var issues []ComplianceIssue

	restrictiveLicenses := []string{"GPL-3.0", "AGPL-3.0", "GPL-2.0", "SSPL"}

	for _, dep := range snapshot.Dependencies {
		// Check for restrictive licenses in production
		if dep.Type == "production" && dep.License != "" {
			for _, restricted := range restrictiveLicenses {
				if dep.License == restricted {
					issues = append(issues, ComplianceIssue{
						Dependency:       dep.Name,
						Version:          dep.Version,
						Issue:            "license",
						Severity:         "high",
						Description:      fmt.Sprintf("Uses restrictive license: %s", dep.License),
						Recommendation:   "Consider replacing with an MIT/Apache-2.0 licensed alternative",
						AutoFixAvailable: false,
					})
				}
			}
		}

		// Check for stale dependencies
		if dep.UpdatedAt != "" {
			updatedTime, _ := time.Parse(time.RFC3339, dep.UpdatedAt)
			twoYearsAgo := time.Now().AddDate(-2, 0, 0)
			if updatedTime.Before(twoYearsAgo) {
				issues = append(issues, ComplianceIssue{
					Dependency:       dep.Name,
					Version:          dep.Version,
					Issue:            "outdated",
					Severity:         "low",
					Description:      "Not updated in over 2 years",
					Recommendation:   "Check if a newer version is available",
					AutoFixAvailable: true,
				})
			}
		}
	}

	return issues, nil
}

// GenerateRevertCommand generates a command to revert a dependency change
func GenerateRevertCommand(change DependencyChange) string {
	switch change.Type {
	case "added":
		return fmt.Sprintf("composer remove %s", change.Name)
	case "removed":
		return fmt.Sprintf("composer require %s:%s", change.Name, change.OldVersion)
	case "updated":
		return fmt.Sprintf("composer require %s:%s", change.Name, change.OldVersion)
	default:
		return ""
	}
}

File: dpb-mcp-rust/src/analyzer/tracker.rs
//! Dependency Tracker - Timestamps and versioning for dependency changes
//! Enables reverting or replacing non-compliant dependencies

use anyhow::Result;
use chrono::{DateTime, Duration, Utc};
use serde::{Deserialize, Serialize};
use sha2::{Sha256, Digest};
use std::collections::HashMap;
use std::fs;
use std::path::Path;

const TRACKER_FILE: &str = ".dpb-dependency-tracker.json";

/// Snapshot of all dependencies at a point in time
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DependencySnapshot {
    pub timestamp: String,
    pub checksum: String,
    pub dependencies: Vec<TrackedDependency>,
    pub metadata: SnapshotMetadata,
}

/// A single tracked dependency with timestamps
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TrackedDependency {
    pub name: String,
    pub version: String,
    #[serde(rename = "type")]
    pub dep_type: String, // "production" or "development"
    #[serde(skip_serializing_if = "Option::is_none")]
    pub added_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub updated_at: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub license: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub security_status: Option<String>,
}

/// Metadata about the snapshot
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SnapshotMetadata {
    pub repo_path: String,
    pub package_manager: String,
    pub total_count: usize,
}

/// A change between two snapshots
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DependencyChange {
    #[serde(rename = "type")]
    pub change_type: String, // "added", "removed", "updated"
    pub name: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub old_version: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub new_version: Option<String>,
    pub timestamp: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
}

/// A compliance issue with a dependency
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ComplianceIssue {
    pub dependency: String,
    pub version: String,
    pub issue: String, // "license", "security", "outdated", "deprecated"
    pub severity: String, // "critical", "high", "medium", "low"
    pub description: String,
    pub recommendation: String,
    pub auto_fix_available: bool,
}

/// Dependency history with categorization
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DependencyHistory {
    pub current_snapshot: DependencySnapshot,
    pub recently_added: Vec<TrackedDependency>,
    pub recently_updated: Vec<TrackedDependency>,
    pub stale: Vec<TrackedDependency>,
}

/// Create a snapshot of current dependencies
pub fn create_dependency_snapshot(repo_path: &str) -> Result<DependencySnapshot> {
    let deps = super::analyze_dependencies(repo_path)?;
    let now = Utc::now().to_rfc3339();
    
    // Load existing tracker to preserve timestamps
    let existing = load_tracker(repo_path).ok();
    let existing_deps: HashMap<String, TrackedDependency> = existing
        .as_ref()
        .map(|s| s.dependencies.iter().map(|d| (d.name.clone(), d.clone())).collect())
        .unwrap_or_default();
    
    let mut tracked: Vec<TrackedDependency> = Vec::new();
    
    for pkg in &deps.tree {
        let existing_dep = existing_deps.get(&pkg.name);
        
        let added_at = existing_dep
            .and_then(|e| e.added_at.clone())
            .unwrap_or_else(|| now.clone());
        
        let updated_at = if existing_dep.map(|e| &e.version) != Some(&pkg.version) {
            now.clone()
        } else {
            existing_dep
                .and_then(|e| e.updated_at.clone())
                .unwrap_or_else(|| now.clone())
        };
        
        tracked.push(TrackedDependency {
            name: pkg.name.clone(),
            version: pkg.version.clone(),
            dep_type: pkg.dep_type.clone().unwrap_or_else(|| "production".to_string()),
            added_at: Some(added_at),
            updated_at: Some(updated_at),
            license: pkg.license.clone(),
            security_status: Some("unknown".to_string()),
        });
    }
    
    // Calculate checksum
    let mut names: Vec<String> = tracked
        .iter()
        .map(|d| format!("{}@{}", d.name, d.version))
        .collect();
    names.sort();
    
    let mut hasher = Sha256::new();
    hasher.update(names.join("|"));
    let hash = hasher.finalize();
    let checksum = hex::encode(&hash[..8]);
    
    Ok(DependencySnapshot {
        timestamp: now,
        checksum,
        dependencies: tracked.clone(),
        metadata: SnapshotMetadata {
            repo_path: repo_path.to_string(),
            package_manager: "composer".to_string(),
            total_count: tracked.len(),
        },
    })
}

/// Load existing tracker from file
pub fn load_tracker(repo_path: &str) -> Result<DependencySnapshot> {
    let tracker_path = Path::new(repo_path).join(TRACKER_FILE);
    let content = fs::read_to_string(tracker_path)?;
    let snapshot: DependencySnapshot = serde_json::from_str(&content)?;
    Ok(snapshot)
}

/// Save snapshot to tracker file
pub fn save_snapshot(repo_path: &str, snapshot: &DependencySnapshot) -> Result<()> {
    let tracker_path = Path::new(repo_path).join(TRACKER_FILE);
    let content = serde_json::to_string_pretty(snapshot)?;
    fs::write(tracker_path, content)?;
    Ok(())
}

/// Compare two snapshots and return changes
pub fn compare_snapshots(old: &DependencySnapshot, new: &DependencySnapshot) -> Vec<DependencyChange> {
    let mut changes = Vec::new();
    
    let old_deps: HashMap<&str, &TrackedDependency> = old
        .dependencies
        .iter()
        .map(|d| (d.name.as_str(), d))
        .collect();
    
    let new_deps: HashMap<&str, &TrackedDependency> = new
        .dependencies
        .iter()
        .map(|d| (d.name.as_str(), d))
        .collect();
    
    // Find added and updated
    for (name, new_dep) in &new_deps {
        if let Some(old_dep) = old_deps.get(name) {
            if old_dep.version != new_dep.version {
                changes.push(DependencyChange {
                    change_type: "updated".to_string(),
                    name: name.to_string(),
                    old_version: Some(old_dep.version.clone()),
                    new_version: Some(new_dep.version.clone()),
                    timestamp: new.timestamp.clone(),
                    reason: None,
                });
            }
        } else {
            changes.push(DependencyChange {
                change_type: "added".to_string(),
                name: name.to_string(),
                old_version: None,
                new_version: Some(new_dep.version.clone()),
                timestamp: new.timestamp.clone(),
                reason: None,
            });
        }
    }
    
    // Find removed
    for (name, old_dep) in &old_deps {
        if !new_deps.contains_key(name) {
            changes.push(DependencyChange {
                change_type: "removed".to_string(),
                name: name.to_string(),
                old_version: Some(old_dep.version.clone()),
                new_version: None,
                timestamp: new.timestamp.clone(),
                reason: None,
            });
        }
    }
    
    changes
}

/// Get dependency history with categorization
pub fn get_dependency_history(repo_path: &str) -> Result<DependencyHistory> {
    let snapshot = create_dependency_snapshot(repo_path)?;
    let now = Utc::now();
    let thirty_days_ago = now - Duration::days(30);
    let one_year_ago = now - Duration::days(365);
    
    let mut recently_added = Vec::new();
    let mut recently_updated = Vec::new();
    let mut stale = Vec::new();
    
    for dep in &snapshot.dependencies {
        if let Some(ref added_at) = dep.added_at {
            if let Ok(added_time) = DateTime::parse_from_rfc3339(added_at) {
                if added_time.with_timezone(&Utc) > thirty_days_ago {
                    recently_added.push(dep.clone());
                }
            }
        }
        
        if let Some(ref updated_at) = dep.updated_at {
            if let Ok(updated_time) = DateTime::parse_from_rfc3339(updated_at) {
                let updated_utc = updated_time.with_timezone(&Utc);
                if updated_utc > thirty_days_ago && dep.updated_at != dep.added_at {
                    recently_updated.push(dep.clone());
                }
                if updated_utc < one_year_ago {
                    stale.push(dep.clone());
                }
            }
        }
    }
    
    Ok(DependencyHistory {
        current_snapshot: snapshot,
        recently_added,
        recently_updated,
        stale,
    })
}

/// Check dependencies for compliance issues
pub fn check_compliance(repo_path: &str) -> Result<Vec<ComplianceIssue>> {
    let snapshot = create_dependency_snapshot(repo_path)?;
    let mut issues = Vec::new();
    
    let restrictive_licenses = ["GPL-3.0", "AGPL-3.0", "GPL-2.0", "SSPL"];
    
    for dep in &snapshot.dependencies {
        // Check for restrictive licenses
        if dep.dep_type == "production" {
            if let Some(ref license) = dep.license {
                for restricted in &restrictive_licenses {
                    if license.to_uppercase().contains(&restricted.to_uppercase()) {
                        issues.push(ComplianceIssue {
                            dependency: dep.name.clone(),
                            version: dep.version.clone(),
                            issue: "license".to_string(),
                            severity: "high".to_string(),
                            description: format!("Uses restrictive license: {}", license),
                            recommendation: "Consider replacing with an MIT/Apache-2.0 licensed alternative".to_string(),
                            auto_fix_available: false,
                        });
                    }
                }
            }
        }
        
        // Check for stale dependencies
        if let Some(ref updated_at) = dep.updated_at {
            if let Ok(updated_time) = DateTime::parse_from_rfc3339(updated_at) {
                let two_years_ago = Utc::now() - Duration::days(730);
                if updated_time.with_timezone(&Utc) < two_years_ago {
                    issues.push(ComplianceIssue {
                        dependency: dep.name.clone(),
                        version: dep.version.clone(),
                        issue: "outdated".to_string(),
                        severity: "low".to_string(),
                        description: "Not updated in over 2 years".to_string(),
                        recommendation: "Check if a newer version is available".to_string(),
                        auto_fix_available: true,
                    });
                }
            }
        }
    }
    
    Ok(issues)
}

/// Generate command to revert a dependency change
pub fn generate_revert_command(change: &DependencyChange) -> String {
    match change.change_type.as_str() {
        "added" => format!("composer remove {}", change.name),
        "removed" => format!("composer require {}:{}", change.name, change.old_version.as_deref().unwrap_or("*")),
        "updated" => format!("composer require {}:{}", change.name, change.old_version.as_deref().unwrap_or("*")),
        _ => String::new(),
    }
}
